references/video_classification/train.py

```diff
@@ -0,0 +1,340 @@
+        metric_logger.update(loss=loss.item(), lr=optimizer.param_groups[0]["lr"])
+        metric_logger.meters['acc1'].update(acc1.item(), n=batch_size)
+        metric_logger.meters['acc5'].update(acc5.item(), n=batch_size)
+        metric_logger.meters['img/s'].update(batch_size / (time.time() - start_time))
+        lr_scheduler.step()
+
+
+def evaluate(model, criterion, data_loader, device):
+    model.eval()
+    metric_logger = utils.MetricLogger(delimiter="  ")
+    header = 'Test:'
+    with torch.no_grad():
+        for image, target in metric_logger.log_every(data_loader, 100, header):
+            image = image.to(device, non_blocking=True)
+            target = target.to(device, non_blocking=True)
+            output = model(image)
+            loss = criterion(output, target)
+
+            acc1, acc5 = utils.accuracy(output, target, topk=(1, 5))
+            # FIXME need to take into account that the datasets
+            # could have been padded in distributed setup
+            batch_size = image.shape[0]
+            metric_logger.update(loss=loss.item())
+            metric_logger.meters['acc1'].update(acc1.item(), n=batch_size)
+            metric_logger.meters['acc5'].update(acc5.item(), n=batch_size)
+    # gather the stats from all processes
+    metric_logger.synchronize_between_processes()
+
+    print(' * Acc@1 {top1.global_avg:.3f} Acc@5 {top5.global_avg:.3f}'
+          .format(top1=metric_logger.acc1, top5=metric_logger.acc5))
+    return metric_logger.acc1.global_avg
+
+
+def _get_cache_path(filepath):
+    import hashlib
+    h = hashlib.sha1(filepath.encode()).hexdigest()
+    cache_path = os.path.join("~", ".torch", "vision", "datasets", "imagefolder", h[:10] + ".pt")
+    cache_path = os.path.expanduser(cache_path)
+    return cache_path
+
+
+def main(args):
+    if args.apex:
+        if sys.version_info < (3, 0):
+            raise RuntimeError("Apex currently only supports Python 3. Aborting.")
+        if amp is None:
+            raise RuntimeError("Failed to import apex. Please install apex from https://www.github.com/nvidia/apex "
+                               "to enable mixed-precision training.")
+
+    if args.output_dir:
+        utils.mkdir(args.output_dir)
+
+    utils.init_distributed_mode(args)
+    print(args)
+    print("torch version: ", torch.__version__)
+    print("torchvision version: ", torchvision.__version__)
+
+    device = torch.device(args.device)
+
+    torch.backends.cudnn.benchmark = True
+
+    # Data loading code
+    print("Loading data")
+    traindir = os.path.join(args.data_path, 'train_avi-480p')
+    valdir = os.path.join(args.data_path, 'val_avi-480p')
+    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                     std=[0.229, 0.224, 0.225])
+
+    print("Loading training data")
+    st = time.time()
+    cache_path = _get_cache_path(traindir)
+    dataset = torch.load("/private/home/fmassa/github/vision/kinetics_train.pth")
```

Comment:
Probably Want to remove hardcoded path?

Reply:
Also, do you literally save the dataset object? (that's what I've been doing but I'm not 100% sure if it's ok to do).

Reply:
this is ok to do in general, but it also serializes the transforms.
My initial thinking was to only serialize `VideoClips`, but then I need to let all datasets be constructed by passing a `VideoClips` object, which I'm not sure I want to do.

Thoughts?

Reply:
I agree that passing VideoClips is a pain.
Here is my experimentation workaround:


```
class Ucf101(UCF101):

    def __init__(self, root, annotation_path, frames_per_clip, step_between_clips=1,
                 fold=1, train=True, spatial_transform=None, audio_transform=None):

        super().__init__(root, frames_per_clip, step_between_clips)
        self.audio_transform = audio_transform
        self.spatial_transform = spatial_transform

    def __getitem__(self, idx):
        video, audio, info, video_idx = self.video_clips.get_clip(idx)
        label = self.samples[video_idx][1]

        if self.audio_transform:
            audio = self.audio_transform(audio)
        if self.spatial_transform:
            video = self.spatial_transform(video)

        return video, audio, label, video_idx

    def set_spatial_transform(self, transform):
        self.spatial_transform = transform
    
    def set_audio_transform(self, transform):
        self.audio_transform = transform
```

Not pretty, but allows you to pass the transform even after the dataset is instantiated and then serialized. TBH, I feel like it would be better if we push to get batch transforms done in both vision and audio, bc I feel like if we had this, saving DS as an object wouldn't be (such) a problem.

Reply:
I added some basic caching mechanism, which can be improved after the release.